File: C_plus_plus/Binary_Tree_Level_Order_Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrder(TreeNode *root) {
        queue<pair<TreeNode *, int>> que;
        vector<vector<int> > result;
        vector<int> tmp;
        int level = 0;
        que.push(make_pair(root,0));
        while(!que.empty()){
            pair<TreeNode *, int> p = que.front();
            que.pop();
            if(p.first){
                if(p.second>level){
                    result.push_back(tmp);
                    tmp.clear();
                    level = p.second;
                }
                tmp.push_back(p.first->val);
                que.push(make_pair(p.first->left,p.second+1));
                que.push(make_pair(p.first->right,p.second+1));
            }
        }
        if(tmp.size()) result.push_back(tmp);
        return result;
    }
};

File: java/Binary_Tree_Level_Order_Traversal.java
// Queue method
// Using current level's count and next level's count to record the length of the two level operated.
// Time is O(n), space is O(n).
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> list = new ArrayList<Integer>();
        Queue<TreeNode> que = new LinkedList<TreeNode>();
        if (root == null) return result;
        int count = 1, cur = 0, next = 0;;
        que.offer(root);
        while (que.peek() != null) {
            TreeNode node = que.poll();
            list.add(node.val);
            if (node.left != null) {
                que.offer(node.left); 
                next++;
            }
            if (node.right != null) {
                que.offer(node.right);
                next++;
            }
            if (++cur == count) {
                result.add(list);
                list = new ArrayList<Integer>();
                count = next;
                next = cur = 0;
                
            }
        }
        return result;
    }
}
